src/chai/SharedPtrManager.cpp

```diff
@@ -0,0 +1,626 @@
+
+  // if we are registering a new pointer record for a pointer where there is already
+  // a pointer record, we assume the old record was somehow abandoned by the host
+  // application and trigger an ACTION_FOUND_ABANDONED callback
+  auto found_pointer_record_pair = m_pointer_map.find(pointer);
+  if (found_pointer_record_pair != m_pointer_map.end()) {
+     msp_pointer_record ** found_pointer_record_addr = found_pointer_record_pair->second;
+     if (found_pointer_record_addr != nullptr) {
+
+        msp_pointer_record *foundRecord = *found_pointer_record_addr;
+        // if it's actually the same pointer record, then we're OK. If it's a different
+        // one, delete the old one.
+        if (foundRecord != record) {
+           CHAI_LOG(Warning, "SharedPtrManager::registerPointer found a record for " <<
+                      pointer << " already there.  Deleting abandoned pointer record.");
+
+           //callback(foundRecord, ACTION_FOUND_ABANDONED, space);
+
+           for (int fspace = CPU; fspace < NUM_EXECUTION_SPACES; ++fspace) {
+              foundRecord->m_pointers[fspace] = nullptr;
+           }
+
+           delete foundRecord;
+        }
+     }
+  }
+
+  CHAI_LOG(Debug, "Registering " << pointer << " in space " << space);
+
+  m_pointer_map.insert(pointer, record);
+
+  for (int i = 0; i < NUM_EXECUTION_SPACES; i++) {
+    if (!record->m_pointers[i]) record->m_owned[i] = true;
+  }
+  record->m_owned[space] = owned;
+
+  if (pointer) {
+     // if umpire already knows about this pointer, we want to make sure its records and ours
+     // are consistent
+     if (m_resource_manager.hasAllocator(pointer)) {
+     //    umpire::util::AllocationRecord *allocation_record = const_cast<umpire::util::AllocationRecord *>(m_resource_manager.findAllocationRecord(pointer));
+     //    //allocation_record->size = record->m_size;
+     }
+     // register with umpire if it's not there so that umpire can perform data migrations
+     else {
+        umpire::util::AllocationRecord new_allocation_record;
+        new_allocation_record.ptr = pointer;
+        //new_allocation_record.size = record->m_size;
+        new_allocation_record.strategy = m_resource_manager.getAllocator(record->m_allocators[space]).getAllocationStrategy();
+
+        m_resource_manager.registerAllocation(pointer, new_allocation_record);
+     }
+  }
+}
+
+void SharedPtrManager::deregisterPointer(msp_pointer_record* record, bool deregisterFromUmpire)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  for (int i = 0; i < NUM_EXECUTION_SPACES; i++) {
+    void * pointer = record->m_pointers[i];
+    if (pointer) {
+       if (deregisterFromUmpire) {
+          m_resource_manager.deregisterAllocation(pointer);
+       }
+       CHAI_LOG(Debug, "De-registering " << pointer);
+       m_pointer_map.erase(pointer);
+    }
+  }
+  if (record != &s_null_record) {
+     delete record;
+  }
+}
+
+//void * SharedPtrManager::frontOfAllocation(void * pointer) {
```

Comment:
Remove and document in an issue. This method might not make sense since with dynamic_casting you may not always get the front of an allocation.